File: src/main/java/com/kingmang/lazurite/runtime/values/LzrArray.java
package com.kingmang.lazurite.runtime.values;

import com.kingmang.lazurite.core.Converters;
import com.kingmang.lazurite.core.Types;
import com.kingmang.lazurite.exceptions.LzrException;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;


public class LzrArray implements LzrValue, Iterable<LzrValue> {

    public static LzrArray of(byte[] array) {
        final int size = array.length;
        final LzrArray result = new LzrArray(size);
        for (int i = 0; i < size; i++) {
            result.set(i, LzrNumber.of(array[i]));
        }
        return result;
    }

    public static LzrArray of(String[] array) {
        final int size = array.length;
        final LzrArray result = new LzrArray(size);
        for (int i = 0; i < size; i++) {
            result.set(i, new LzrString(array[i]));
        }
        return result;
    }
    
    public static LzrArray add(LzrArray array, LzrValue value) {
        final int last = array.elements.length;
        final LzrArray result = new LzrArray(last + 1);
        System.arraycopy(array.elements, 0, result.elements, 0, last);
        result.elements[last] = value;
        return result;
    }
    
    public static LzrArray merge(LzrArray array1, LzrArray array2) {
        final int length1 = array1.elements.length;
        final int length2 = array2.elements.length;
        final int length = length1 + length2;
        final LzrArray result = new LzrArray(length);
        System.arraycopy(array1.elements, 0, result.elements, 0, length1);
        System.arraycopy(array2.elements, 0, result.elements, length1, length2);
        return result;
    }
    
    public static LzrString joinToString(LzrArray array, String delimiter, String prefix, String suffix) {
        final StringBuilder sb = new StringBuilder();
        for (LzrValue value : array) {
            if (!sb.isEmpty()) sb.append(delimiter);
            else sb.append(prefix);
            sb.append(value.asString());
        }
        sb.append(suffix);
        return new LzrString(sb.toString());
    }
    
    private final LzrValue[] elements;

    public LzrArray(int size) {
        this.elements = new LzrValue[size];
    }

    public LzrArray(LzrValue[] elements) {
        this.elements = new LzrValue[elements.length];
        System.arraycopy(elements, 0, this.elements, 0, elements.length);
    }
    
    public LzrArray(List<LzrValue> values) {
        final int size = values.size();
        this.elements = values.toArray(new LzrValue[size]);
    }
    
    public LzrArray(LzrArray array) {
        this(array.elements);
    }

    public LzrValue[] getCopyElements() {
        final LzrValue[] result = new LzrValue[elements.length];
        System.arraycopy(elements, 0, result, 0, elements.length);
        return result;
    }
    
    @Override
    public int type() {
        return Types.ARRAY;
    }

    public int size() {
        return elements.length;
    }

    public LzrValue get(int index) {
        return elements[index];
    }

    public LzrValue get(LzrValue index) {
        return switch (index.asString()) {
            case "length" -> LzrNumber.of(size());
            case "isEmpty" -> Converters.voidToBoolean(() -> size() == 0);
            default -> get(index.asInt());
        };
    }


    public void set(int index, LzrValue value) {
        elements[index] = value;
    }

    @NotNull
    @Override
    public LzrValue[] raw() {
        return elements;
    }

    @Override
    public int asInt() {
        throw new LzrException("TypeExeption","Cannot cast array to integer");
    }

    @Override
    public double asNumber() {
        throw new LzrException("TypeExeption","Cannot cast array to number");
    }

    @NotNull
    @Override
    public String asString() {
        return Arrays.toString(elements);
    }

    @NotNull
    @Override
    public int[] asArray() {
        return new int[0];
    }


    @NotNull
    @Override
    public Iterator<LzrValue> iterator() {
        return Arrays.asList(elements).iterator();
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 79 * hash + Arrays.deepHashCode(this.elements);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null) return false;
        if (getClass() != obj.getClass())
            return false;
        final LzrArray other = (LzrArray) obj;
        return Arrays.deepEquals(this.elements, other.elements);
    }

    @Override
    public int compareTo(@NotNull LzrValue o) {
        if (o.type() == Types.ARRAY) {
            final int lengthCompare = Integer.compare(size(), ((LzrArray) o).size());
            if (lengthCompare != 0) return lengthCompare;
        }
        return asString().compareTo(o.asString());
    }

    @Override
    public String toString() {
        return asString();
    }
}

File: src/main/java/com/kingmang/lazurite/runtime/values/LzrArray.kt
package com.kingmang.lazurite.runtime.values

import com.kingmang.lazurite.core.Converters
import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.core.asLzrArrayOrNull
import com.kingmang.lazurite.core.throwTypeCastException

open class LzrArray(private val elements: Array<LzrValue>, unit: Unit) : LzrValue, Iterable<LzrValue> {
    constructor(size: Int) :
            this(Array(size) { LzrNull }, Unit)
    constructor(size: Int, init: (index: Int) -> LzrValue) :
            this(Array(size, init), Unit)
    constructor(values: Array<LzrValue>) :
            this(values.copyOf(), Unit)
    constructor(values: List<LzrValue>) :
            this(values.toTypedArray(), Unit)
    constructor(array: LzrArray) :
            this(array.elements.copyOf(), Unit)

    open val copyElements: Array<LzrValue>
        get() = this.elements.copyOf()

    override fun type(): Int =
        Types.ARRAY

    open fun size(): Int =
        this.elements.size

    open fun isEmpty(): Boolean =
        this.size() == 0

    open fun add(value: LzrValue): LzrArray =
        add(this, value)

    open fun merge(array: LzrArray): LzrArray =
        merge(this, array)

    open operator fun get(index: Int): LzrValue =
        this.elements[index]

    operator fun get(index: LzrValue): LzrValue =
        when (index.asString()) {
            "length" -> LzrNumber.of(this.size())
            "isEmpty" -> Converters.voidToBoolean { this.size() == 0 }
            else -> this[index.asInt()]
        }

    open operator fun set(index: Int, value: LzrValue) {
        this.elements[index] = value
    }

    override fun raw(): Array<LzrValue> =
        this.elements

    override fun asInt(): Int =
        throwTypeCastException("array", "integer")

    override fun asNumber(): Double =
        throwTypeCastException("array", "number")

    override fun asString(): String =
        this.elements.contentToString()

    override fun asArray(): IntArray =
        IntArray(0)

    override fun iterator(): Iterator<LzrValue> =
        this.elements.iterator()

    override fun hashCode(): Int {
        var hash = 5
        hash = 79 * hash + this.elements.contentDeepHashCode()
        return hash
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null || this.javaClass != other.javaClass)
            return false
        other as LzrArray
        return elements.contentDeepEquals(other.elements)
    }

    override fun compareTo(other: LzrValue): Int {
        other.asLzrArrayOrNull()?.let {
            val lengthCompare = this.size().compareTo(it.size())
            if (lengthCompare != 0) {
                return lengthCompare
            }
        }
        return this.asString().compareTo(other.asString())
    }

    override fun toString(): String =
        this.asString()

    companion object {
        @JvmStatic
        fun of(vararg values: LzrValue): LzrArray =
            LzrArray(arrayOf(*values))

        @JvmStatic
        fun of(array: ByteArray): LzrArray =
            LzrArray(array.size) { LzrNumber.of(array[it].toInt()) }

        @JvmStatic
        fun of(array: Array<String>): LzrArray =
            LzrArray(array.size) { LzrString(array[it]) }

        @JvmStatic
        fun add(array: LzrArray, value: LzrValue): LzrArray {
            val last = array.elements.size
            return LzrArray(last + 1).apply {
                array.elements.copyInto(this.elements)
                this.elements[last] = value
            }
        }

        @JvmStatic
        fun merge(array1: LzrArray, array2: LzrArray): LzrArray {
            val length1 = array1.elements.size
            val length2 = array2.elements.size
            val length = length1 + length2
            val result = LzrArray(length)
            array1.elements.copyInto(result.elements, endIndex = length1)
            array2.elements.copyInto(result.elements, destinationOffset = length1, endIndex = length2)
            return result
        }

        fun joinToString(array: LzrArray, delimiter: String = "", prefix: String = "", suffix: String = ""): LzrString {
            val sb = StringBuilder()
            for (value in array) {
                if (sb.isNotEmpty())
                    sb.append(delimiter)
                else sb.append(prefix)
                sb.append(value.asString())
            }
            sb.append(suffix)
            return LzrString(sb.toString())
        }
    }
}
